torch/_higher_order_ops/templated_attention.py

```diff
@@ -0,0 +1,209 @@
+    proxy_mode.tracer.root.register_module("sdpa_score", score_graph)
+    node_args = (query, key, value, score_graph, *other_buffers)
+    proxy_args = pytree.tree_map(proxy_mode.tracer.unwrap_proxy, node_args)
+    out_proxy = proxy_mode.tracer.create_proxy(
+        "call_function", sdpa, proxy_args, {}, name="templated_attention"
+    )
+    return track_tensor_tree(
+        example_out, out_proxy, constant=None, tracer=proxy_mode.tracer
+    )
+
+
+@sdpa.py_impl(ProxyTorchDispatchMode)
+def sdpa_proxy_torch_dispatch_mode(
+    mode: ProxyTorchDispatchMode,
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    score_mod: Callable,
+    *other_buffers: torch.Tensor,
+):
+    assert mode is not None, "Mode should always be enabled for python fallback key"
+    if mode.enable_tracing:
+        return trace_sdpa(mode, query, key, value, score_mod, *other_buffers)
+    else:
+        return sdpa(query, key, value, score_mod, *other_buffers)
+
+
+@sdpa.py_functionalize_impl
+def sdpa_functionalize(
+    ctx: torch._subclasses.functional_tensor.BaseFunctionalizeAPI,
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    score_mod: Callable,
+    *other_buffers: torch.Tensor,
+):
+    """Defines the functionalization rules for the sdpa operator.
+
+    Write now we are unwrapping each tensor and then redispatching to the next, however we want to
+    guard against any mutations or aliases in the score_mod function. To the other_buffers since those
+    are free variables.
+    """
+    query_unwrapped = ctx.unwrap_tensors(query)
+    key_unwrapped = ctx.unwrap_tensors(key)
+    value_unwrapped = ctx.unwrap_tensors(value)
+    other_buffers_unwrapped = ctx.unwrap_tensors(other_buffers)
+    example_vals = [torch.zeros((), dtype=query.dtype)] + [
+        torch.zeros((), dtype=torch.int) for _ in range(4)
+    ]
+    with ctx.redispatch_to_next() as m:
+        functional_score_mod = ctx.functionalize(score_mod)
+        pre_dispatch = hasattr(ctx, "mode") and ctx.mode.pre_dispatch
+        mutates = _has_potential_branch_input_mutation(
+            functional_score_mod, example_vals, pre_dispatch
+        )
+        # The only care about mutations of existing buffers since we can't replay these.
+        # However, we can just error if anything is detected
+        if mutates:
+            raise UnsupportedAliasMutationException("Mutations detected in score_mod")
+
+        out = sdpa(
+            query_unwrapped,
+            key_unwrapped,
+            value_unwrapped,
+            functional_score_mod,
+            *other_buffers_unwrapped,
+        )
+    return ctx.wrap_tensors(out)
+
+
+@sdpa.py_impl(FakeTensorMode)
+def sdpa_fake_tensor_mode(
+    mode: FakeTensorMode,
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    score_mod: Callable,
+    *other_buffers: torch.Tensor,
+):
+    with mode:
```

Comment:
@zou3519 When calling into the vmap impl under fake mode I get the error I pasted in the comment. If I just return the output everything seems to work, @bdhirsh  suggested trying this, I am still not entirely sure why

Reply:
Does this work?

Reply:
cool!